lib/rust/parser/debug/src/bin/bench_parse.rs

```diff
@@ -11,24 +12,78 @@
 
 
 
-// =============
-// === Tests ===
-// =============
+// ===========
+// === CLI ===
+// ===========
 
 fn main() {
     let args = std::env::args().skip(1);
     let parser = enso_parser::Parser::new();
     let parse_time: std::time::Duration = args
         .map(|path| {
-            let code = std::fs::read_to_string(path).unwrap();
-            let mut code = code.as_str();
-            if let Some((_meta, code_)) = enso_parser::metadata::parse(code) {
-                code = code_;
-            }
+            let code = read_source(path).unwrap();
             let start = std::time::Instant::now();
-            std::hint::black_box(parser.run(code));
+            std::hint::black_box(parser.run(&code));
             start.elapsed()
         })
         .sum();
     println!("Parse time: {} ms", parse_time.as_millis());
 }
+
+fn read_source(path: impl AsRef<Path>) -> io::Result<String> {
+    let code = fs::read_to_string(path)?;
+    Ok(if let Some((_meta, code)) = enso_parser::metadata::parse(&code) {
+        code.to_owned()
+    } else {
+        code
+    })
+}
+
+
+
+// ===============================
+// === `cargo bench` interface ===
+// ===============================
+
+extern crate test;
+
+use std::fs::DirEntry;
+use std::fs::{self};
+use std::io;
+use std::path::Path;
+
+fn visit_files<F: FnMut(&DirEntry)>(dir: &Path, f: &mut F) -> io::Result<()> {
+    if dir.is_dir() {
+        for entry in fs::read_dir(dir)? {
+            let entry = entry?;
+            let path = entry.path();
+            if path.is_dir() {
+                visit_files(&path, f)?;
+            } else {
+                f(&entry);
+            }
+        }
+    }
+    Ok(())
+}
+
+#[bench]
+fn bench_std_lib(b: &mut test::Bencher) {
```

Comment:
Note that there is a framework for benchmarks in the engine. We currently have benchmarks for the engine compiler, like [InlineCOmpilerBenchmark](https://github.com/enso-org/enso/blob/develop/engine/runtime-benchmarks/src/main/java/org/enso/compiler/benchmarks/inline/InlineCompilerBenchmark.java). We don't have benchmarks for parser yet, but they should not be difficult to implement.

Is this benchmark run automatically on the CI, or is it supposed to be run manually? Shouldn't we just add parser benchmarks to the already existing engine [runtime-benchmarks](https://github.com/enso-org/enso/tree/develop/engine/runtime-benchmarks/src/main) infrastructure?

Reply:
It would be great to have an `EnsoParser` benchmark to measure the overhead of JNI calls when doing `TreeToIr`. Probably just a copy of [ManyLocalVarsBenchmark](https://github.com/enso-org/enso/blob/develop/engine/runtime-benchmarks/src/main/java/org/enso/compiler/benchmarks/module/ManyLocalVarsBenchmark.java) that doesn't call compiler, but only `TreeToIr`.